### PR TITLE
GPT-1

### PR DESCRIPTION
**Pull request by workofart**
## Description
Implementing the Paper: [Improving Language Understanding by Generative Pre-Training](https://www.semanticscholar.org/paper/Improving-Language-Understanding-by-Generative-Radford-Narasimhan/cd18800a0fe0b668a1cc19f2ec95b5003d0a5035) (a.k.a. GPT-1)

Leveraging some reusable modules (e.g. `MultiHeadAttention`) from the `example/transformers.py` .
This is a decoder-only architecture, so the decoder module needed to be created from scratch based on the paper.

Other changes:
- Added ModuleList module to support nested list of layers in another Module. Without this, the top-level model (e.g. GPT-1) would not be aware of sub-layers/modules (e.g. MultiHeadedAttention) nested as attributes in the top-level model and training would not update those parameters in the sub-layers.
- Add teacher enforcing inference to measure how well the model is predicting the next token by feeding in the groundtruth sequence as context. This is used to quickly see if the model can overfit nicely.
- Refactored some utility functions between the `gpt-1.py` and `transformers.py` files.

## Training and Sample Evaluation Results
TBD

### CODE REVIEW

File: autograd/tools/data.py [hunk 1]
```diff
@@ -94,6 +94,7 @@ def __init__(
         self.batches_y = []
         self.source_masks = []
         self.target_masks = []
+        self.causal_mask = []
```
Comment (workofart) [inline autograd/tools/data.py:97]: In GPT-1, we only apply causal masking to prevent peeking into the future tokens when training, we are feeding the GPT-1 model chunks of text in equal lengths so no padding (or masking) is needed. This is different from the original Transformers architecture where we needed to pad the input text with varying lengths.
